### ISSUE TITLE
Library wrapping Ava

### ISSUE THREAD

**Issue by ehmicky** [avajs/eslint-plugin-ava#268]
Some libraries might be wrapping Ava, for example:

```js
const test = require('custom-ava')
```

or:

```js
const test = require('custom-ava')(require('ava'))
```

Also the `test()` function (and its variants `.skip(...)`, etc.) might be wrapped as well and provide slightly different arguments:

```js
const test = require('ava-test-each')
test(
  ['red', 'green', 'blue'], 
  'example test', 
  (t, color) => {
    t.is(typeof color, 'string')
  }
)
```

I think this does not work with `eslint-plugin-ava` because this plugin might rely on (correct me if I'm wrong):
  - the library exporting `test()` being called `"ava"`
  - the `test()` signature not being modified
  - the main test function being called `test()` (that one should not be an issue though)

I am considering developing a small wrapper around Ava but I am realizing using that wrapper might not be compatible with `eslint-plugin-ava`. Would it be possible to add support for such wrappers?

I understand that might add lots of complexity, so feel free to close this issue if you don't think it's a good idea :)

**Comment by sindresorhus** [comment]
If they wrapped instance might provide other arguments, there are lots of rule assumptions that go out the window.

**Comment by sindresorhus** [comment]
I feel like instead of wrapping AVA, AVA should provide hooks that enable new functionality. Can you describe the problems you’re trying to solve by wrapping AVA?

**Comment by ehmicky** [comment]
I tend to iterate tests over different inputs (as opposed to create several tests). I love Ava but I personally find the syntax for macros not very intuitive and would prefer a syntax closer to [Jest's `test.each()`](https://jestjs.io/docs/en/api#testeachtable-name-fn-timeout) and `Array#forEach()`.

Besides this difference of syntax, there are additional features I would like:
  - iterating over a cartesian product of inputs. This can be achieved with Ava by using several `Array#forEach()` but it's a little verbose.
  - using the stringified inputs as title instead of specifying a test title for each iteration. The generated title should work with any JavaScript type, be always unique (by appending a duplication counter) and be truncated when too long. 

I have created a library [`test-each`](https://github.com/ehmicky/test-each) that does just that: 

```js
const test = require('ava')
const { each } = require('test-each')

// The code we are testing
const multiply = require('./multiply.js')

// Run this test 4 times using every possible combination of inputs
each([0.5, 10], [2.5, 5], ({ title }, first, second) => {
  test(`should mix integers and floats | ${title}`, t => {
    t.is(typeof multiply(first, second), 'number')
  })
})
```

The library is designed to work with any test runner. However I want to add an adapter for Ava to make the syntax even simpler:

```js
test([0.5, 10], [2.5, 5], 'should mix integers and floats', (t, first, second) => {
  t.is(typeof multiply(first, second), 'number')
})
```
